Add option to use OS specific colour theme

shafinkamal opened the pull request:
Resolves #8729.

Currently, I have added a new Radio Button (see below) which the user would select so that their JabRef theme corresponds to the theme that their OS is currently. I have achieved that by referencing the [jSystemThemeDetector library](https://github.com/Dansoftowner/jSystemThemeDetector)

<img width="1440" alt="Screen Shot 2022-10-18 at 8 19 37 pm" src="https://user-images.githubusercontent.com/111040413/196390526-2f0c6c84-cfd6-458e-a40e-156b3cdfb9bb.png">

However, I am currently at a roadblock and would greatly appreciate some help. If my MacOS Theme is 'Light', I can change it to 'Dark' and within JabRef if I select 'Automatic detection' and then 'Save', JabRef will change from Light to Dark. However when going back the JabRef Preferences, the setting has not stayed at 'Automatic detection' and instead has moved to 'Dark theme'. This means that next time I change from Dark to Light on my MacOS System Preferences, this change won't be reflected in JabRef. I've attached a demo to visually show the behaviour.

![jabref theme demo gif](https://user-images.githubusercontent.com/111040413/196399610-29f31462-598e-4419-abb6-8cb9768b7937.gif)

Any pointers please? Thank you!!!

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

Siedlerchr commented on src/main/java/org/jabref/preferences/JabRefPreferences.java:
```diff
@@ -2067,7 +2067,8 @@ public AppearancePreferences getAppearancePreferences() {
                 getInt(MAIN_FONT_SIZE),
-                new Theme(get(FX_THEME))
+                new Theme(get(FX_THEME)),
+                false
```
You will need to add a new constant for the new flag and get the value here : sth like getBoolean(AUTOMATIC_THEME_DETECTION)  and then add a new listener so that the value gets stored and read on change

ThiloteE commented:
To ease organizational workflows I have linked the pull-request to the issue with syntax as described in https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

> <h2 id="linking-a-pull-request-to-an-issue-using-a-keyword">Linking a pull request to an issue using a keyword</h2>
> <p>You can link a pull request to an issue by using a supported keyword 
> in the pull request's description or in a commit message. The pull 
> request <strong>must be</strong> on the default branch.</p>
> <ul>
> <li>close</li>
> <li>closes</li>
> <li>closed</li>
> <li>fix</li>
> <li>fixes</li>
> <li>fixed</li>
> <li>resolve</li>
> <li>resolves</li>
> <li>resolved</li>
> </ul>
> <p>If you use a keyword to reference a pull request comment in another 
> pull request, the pull requests will be linked. Merging the referencing 
> pull request also closes the referenced pull request.</p>
> <p>The syntax for closing keywords depends on whether the issue is in the same repository as the pull request.</p>

shafinkamal commented:
Updated behaviour.

![Jabref theme switch demo](https://user-images.githubusercontent.com/111040413/196663374-bc5486b2-3c72-4417-8739-b9160b62661c.gif)

shafinkamal commented:
I've just noticed that I (at some point) pulled from upstream when in my remote branch accidentally which is why there are other people's contributions in this PR. Will this cause a problem? If so, let me know and I'll fix it up tomorrow.

Thanks (sorry about that)

Siedlerchr commented:
@shafinkamal  That's no problem, but you will probably need to sync your fork's main branch so that github correclty detecs the changes again 

<img width="1005" alt="grafik" src="https://user-images.githubusercontent.com/320228/196772945-e478f821-7054-48a2-9964-106bdcf17453.png">

Thanks for your PR so far, it's overall looking good. But unfortunately the new dependency is blocked by https://bugs.openjdk.org/browse/JDK-8240567

shafinkamal commented:
> ... unfortunately the new dependency is blocked by https://bugs.openjdk.org/browse/JDK-8240567

I have now removed the jSystemThemeDetector dependency and modified the theme detection (see `SystemThemeDetector`). This change however does mean that the detector won't listen to changes of the theme. Perhaps Nevertheless, the underlying implementation is working 👍🏽.

`SystemThemeDetector` class is extracted from this [GitHub Gist](https://gist.github.com/HanSolo/7cf10b86efff8ca2845bf5ec2dd0fe1d).

Further information on it is found [here](https://harmoniccode.blogspot.com/2021/02/poor-mans-dark-mode-detection.html).

Siedlerchr commented:
Thanks for the integration! I think you now need to adjust the failing architecture test because of the usage of the standards streams class

shafinkamal commented:
Anything I can do to fix the failing tests or is it all good on my end? the fails seem to be related to servers and what not...

Anyways, let me know. 

Thanks!

Siedlerchr commented:
Thanks it looks good, the failing tests are related to some jitpack error not resolving the dependency correctly.

calixtus commented:
U sure this works with modern verions of windows too, because it looks like microsoft removed some registry keys:
https://learn.microsoft.com/en-us/answers/questions/715081/how-to-detect-windows-dark-mode.html
?

shafinkamal commented:
I'm not too sure about the modern versions of Windows as I have geared this towards MacOS, because the issue (#8729) is specific to MacOS and I have MacOS so I can manually test it on modern versions of that.

I had included some Windows operations for the sake of coverage but perhaps another issue can be made to ensure further coverage of Windows (and other OS's).

If we're worried about the code breaking the application on the more recent versions of Windows, we can comment out the Windows section and make this a purely MacOS specific feature (for the time being).

calixtus commented on src/main/java/org/jabref/gui/theme/SystemThemeDetector.java:
```diff
@@ -0,0 +1,121 @@
+    private static final String REGDWORD_TOKEN = "REG_DWORD";
+    private static final String DARK_THEME_CMD = REGQUERY_UTIL + "\"HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize\"" + " /v AppsUseLightTheme";
+
+    public final boolean isDarkMode() {
+        switch (getOperatingSystem()) {
+            case WINDOWS: return isWindowsDarkMode();
+            case MACOS: return isMacOsDarkMode();
+            // currently theme detection not implemented for Linux and Solaris
+            case LINUX: return false;
+            case SOLARIS: return false;
+            default: return false;
+        }
+    }
```
Can you please split up this class and use the existing OS package and classes to provide a method isDarkMode instead of introducing some code duplication to test for which os is runnning here?

calixtus commented on src/main/java/org/jabref/gui/theme/SystemThemeDetector.java:
```diff
@@ -0,0 +1,121 @@
+    @AllowedToUseStandardStreams("")
+    public static boolean isWindowsDarkMode() {
+        try {
+            Process process = Runtime.getRuntime().exec(DARK_THEME_CMD);
+            StreamReader reader = new StreamReader(process.getInputStream());
```
You really shoud you JNA here, we already use it in other occasions and its a way cleaner solution.

calixtus commented on src/main/java/org/jabref/preferences/AppearancePreferences.java:
```diff
@@ -13,11 +13,22 @@ public class AppearancePreferences {
         this.shouldOverrideDefaultFontSize = new SimpleBooleanProperty(shouldOverrideDefaultFontSize);
         this.mainFontSize = new SimpleIntegerProperty(mainFontSize);
         this.theme = new SimpleObjectProperty<>(theme);
+        this.automaticThemeDetectionFlag = new SimpleBooleanProperty(automaticThemeDetectionFlag);
```
(should)AutoDetectTheme(Property)

calixtus commented on src/main/java/org/jabref/gui/preferences/appearance/AppearanceTab.fxml:
```diff
@@ -47,4 +49,9 @@
             </tooltip>
         </Button>
     </HBox>
+    <Label fx:id="restartJabRefHint" styleClass="text" text="${'\nRestarting JabRef may be required to display changes in theme preferences.'}">
+        <font>
+            <Font name="Italic" size="13.0"/>
+        </font>
+    </Label>
```
Why just fixed the restart for theme change issue a few months back. This would be a major regression.

calixtus commented:
Thanks for your interest in JabRef programming. I like the idea, yet there are some issues with you changes that should be adressed...

calixtus commented:
![grafik](https://user-images.githubusercontent.com/50491877/197291649-9023e6d0-f60c-4a4e-a488-900bb4e572d0.png)
Make the auto detect option a checkbox and grey out the other options, when checked.

shafinkamal commented:
Hi @calixtus,

Thanks for the feedback. I'm quite busy this weekend and the following week, but I'll get around to implementing these changes when I have some more free time.

In the meantime, could you please assign me to issue #8729?

Thanks!

HoussemNasri commented:
Hey @shafinkamal. Do you plan to continue working on this?

Siedlerchr commented:
Closing this issue due to inactivity :zzz: 
Please reopen the issue with additional information if the problem persists.
